File: 327-multiplyForBigNumbers.py
# given two big numbers as string,return the multiply results

def  multiply(a,b):
    n,m=len(a),len(b)
    ans=[0]*(n+m)
    a=a[::-1]
    b=b[::-1]
    
    for i  in range(n):
        for j in range(m):
            ans[i+j]+=(int)(a[i])*(int)(b[j])
            ans[i+j+1]+=ans[i+j]//10
            ans[i+j]%=10
            
    return ("".join(map(str,ans[::-1]))).lstrip("0")
a="123"
b="456"

print(multiply(a,b))

File: 327-multiplyForBigNumbers.cpp
//given two big numbers as string format,calculate the multiply 

#include<string>
#include<vector>
#include<iostream>
#include<algorithm>  //for std::reverse
using DT=std::string;

void bigNumbersMultiply(DT& sa, DT& sb, DT& res)  //can not be const DT&, as we will reverse the source strings
{
    int lenA=sa.size(),lenB=sb.size();
    if(!lenA || !lenB)
        return ;

    std::vector<int> tmp(lenA+lenB,0);
    std::reverse(sa.begin(),sa.end());
    std::reverse(sb.begin(),sb.end());
    for(int i=0;i<lenA;++i)
    {
        for(int j=0;j<lenB;++j)
        {
            tmp[i+j]+=((sa[i]-'0')*(sb[j]-'0'));
            tmp[i+j+1]+=tmp[i+j]/10;
            tmp[i+j]%=10;
        }
    }
    std::reverse(tmp.begin(),tmp.end());
    int idx=0;
    while((idx<tmp.size())&&(!tmp[idx])) ++idx;
    if(idx==tmp.size())  //here should be tmp.size()
        res="0";   
    for(;idx<tmp.size();++idx)
    {
        res+=std::to_string(tmp[idx]);
    }
}


int main()
{
    DT a{"123"},b{"0"};
    DT res{};
    bigNumbersMultiply(a,b,res);
    std::cout<<res;
}